### PR TITLE
Backward: Add friendly stack trace capability.

### PR DESCRIPTION
**Pull request by dnoe**
Use the Backward-cpp project (https://github.com/bombela/backward-cpp)
to log stack traces on demand as desired for debugging and in response
to serious errors like PANIC() and assertion failures.

Unfortunately the best output from Backward-cpp requires significant
dependencies (elfutils, or worse GPL binutils) that we don't want to
pull in.  Instead, a Python script provides automated de-interleaving of
stack trace output from multiple threads and calls addr2line to decode
the addresses in the stack trace using the DWARF info present in the
binary.  The tool accepts Envoy log output on stdin or takes arguments
for a command to run as a child process and backtrace enhance the
output. For example, this command will run a single test under the
backtrace enhancer and show the output:

bazel test -c dbg //test/common/common:backwards_test
--run_under=`pwd`/tools/stack_decode.py
--strategy=TestRunner=standalone --cache_test_results=no
--test_output=all

Since this is intended only as a debugging tool the test care merely
exercises the code to ensure it does not crash without an affirmative
passing case.

### CODE REVIEW

File: tools/stack_decode.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+# Output one stacktrace after passing it through addr2line with appropriate
+# options
+def output_stacktrace(thread_id, traceinfo):
+  (log_prefix, object_file, stack) = traceinfo
+  piped_input = ""
+  for stack_frame in stack:
+    piped_input += (stack_frame[1] + '\n')
+  addr2line = subprocess.Popen(['addr2line', '-Cpisfe', object_file], stdin=subprocess.PIPE, stdout=subprocess.PIPE)
+  (output_stdout, output_stderr) = addr2line.communicate(piped_input)
```
Comment (htuch) [inline tools/stack_decode.py]: If you're not using `output_stderr`, it might be clearer to write `output_stdout, _ = ...`

File: tools/stack_decode.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+def decode_stacktrace_log(input_source):
+  traces = {}
+  trace_begin_re = re.compile('^(.+)\[backtrace\] Backtrace obj<(.+)> thr<(\d+)')
+  stackaddr_re = re.compile('\[backtrace\] thr<(\d+)> #(\d+) (0x[0-9a-fA-F]+)$')
+  trace_end_re = re.compile('\[backtrace\] end backtrace thread (\d+)')
+
+  # build a dictionary indexed by thread_id:
+  #   value is a tuple (log_prefix, obj_file, [list of addresses])
+  while True:
+    line = input_source.readline()
```
Comment (htuch) [inline tools/stack_decode.py]: How ctrl-c safe is this? If I run a test, then mid-way ctrl-c, killing Bazel and this script, does the child also die, or does it get turned into a zombie? It might be totally fine, I'm just curious as I've written subprocess code that creates zombies in the past this way, so could be good to quickly validate.

File: source/server/server.cc [hunk 3]
```diff
@@ -15,6 +15,7 @@
 #include "envoy/upstream/cluster_manager.h"
 
 #include "common/api/api_impl.h"
+#include "common/common/backward.h"
```
Comment (dnoe) [inline source/server/server.cc]: This was a relic of previous testing.  I was going to do the signal handler in a separate PR, I had a few questions about how to get the conditional option through the build system.

File: tools/stack_decode.py (hunk 2 above)
Comment (dnoe) [inline tools/stack_decode.py]: If I start an Envoy like this:

tools/stack_decode.py bazel-bin/source/exe/envoy-static -c config.json

Then ctrl-C does the right thing.  It does the expected thing when the output is piped into stack_decode.py too.  I can add a catch for KeyboardInterrupt to avoid the ugly stack trace, though...

File: tools/stack_decode.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    if line == '':
+      return # EOF
+    begin_trace_match = trace_begin_re.search(line)
+    if begin_trace_match:
+      (log_prefix, objfile, thread_id) = begin_trace_match.groups()
```
Comment (dnoe) [inline tools/stack_decode.py]: I defer to your preference on this, my Python experience is mostly at the hack level.

File: tools/stack_decode.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+def output_stacktrace(thread_id, traceinfo):
+  (log_prefix, object_file, stack) = traceinfo
+  piped_input = ""
+  for stack_frame in stack:
+    piped_input += (stack_frame[1] + '\n')
+  addr2line = subprocess.Popen(['addr2line', '-Cpisfe', object_file], stdin=subprocess.PIPE, stdout=subprocess.PIPE)
+  (output_stdout, output_stderr) = addr2line.communicate(piped_input)
+  output_lines = output_stdout.split('\n')
+
+  resolved_stack_frames = zip(range(1, len(output_lines)), output_lines)
```
Comment (dnoe) [inline tools/stack_decode.py]: Familiar story :)

File: tools/stack_decode.py (hunk 1 above)
Comment (dnoe) [inline tools/stack_decode.py]: Another Python feature I didn't know about, cool!

File: tools/stack_decode.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+      return # EOF
+    begin_trace_match = trace_begin_re.search(line)
+    if begin_trace_match:
+      (log_prefix, objfile, thread_id) = begin_trace_match.groups()
+      traces[thread_id] = (log_prefix, objfile, [])
+      continue
+    match = stackaddr_re.search(line)
+    if match:
+      (thread_id, frame_idx, address) = match.groups()
+      traces[thread_id][2].append((frame_idx, address))
```
Comment (dnoe) [inline tools/stack_decode.py]: Today (well last night) I learned about namedtuple.  Useful!

File: source/common/common/backward.h [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+   */
+  void Log() {
+#ifdef NDEBUG
+    if (!log_in_prod_) {
+      return;
+    }
+#endif
+    backward::TraceResolver resolver;
+    resolver.load_stacktrace(stack_trace_);
+    // Our own call to Capture() should generate at least 3 stack frames, so we
```
Comment (dnoe) [inline source/common/common/backward.h]: It might not be - and will also be untrue if I do the signal handler implementation as I realized last night.  I'll update this.

Comment (dnoe) [inline source/common/common/backward.h]: Related to this, there isn't actually any way in the current set of build options to achieve optimizations plus -g, and without -g there isn't really any useful information in the backtrace.  Is -g really a runtime performance issue? It increases the size of the binary but the DWARF symbols should never be loaded unless they are accessed by something like gdb or addr2line.

Comment (dnoe) [inline source/common/common/backward.h]: Nevermind, there is --define debug_symbols=yes.  But it seems like doing it this way eliminates useful debug information for production.

Comment (htuch) [inline source/common/common/backward.h]: I don't mind making this the default for `-c opt` and override Bazel's opinion on this. @mattklein123 thoughts?

File: tools/stack_decode.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+# the entire stack trace has been read.  End when EOF received.
+def decode_stacktrace_log(input_source):
+  traces = {}
+  trace_begin_re = re.compile('^(.+)\[backtrace\] Backtrace obj<(.+)> thr<(\d+)')
+  stackaddr_re = re.compile('\[backtrace\] thr<(\d+)> #(\d+) (0x[0-9a-fA-F]+)$')
+  trace_end_re = re.compile('\[backtrace\] end backtrace thread (\d+)')
+
+  # build a dictionary indexed by thread_id:
+  #   value is a namedtuple (log_prefix, obj_file, [list of addresses])
+  Backtrace = collections.namedtuple('Backtrace', 'log_prefix obj_file, address_list')
```
Comment (htuch) [inline tools/stack_decode.py]: General preference for declaring types that are used by multiple functions at the outer scope. You don't need to destruct the traceinfo in `output_stacktrace` if you do this.

File: tools/stack_decode.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+  try:
+    while True:
+      line = input_source.readline()
+      if line == '':
+        return # EOF
+      begin_trace_match = trace_begin_re.search(line)
+      if begin_trace_match:
+        log_prefix, objfile, thread_id = begin_trace_match.groups()
+        traces[thread_id] = Backtrace(log_prefix = log_prefix,
```
Comment (htuch) [inline tools/stack_decode.py]: Can you run the Google pyformat on this file? I think it prefers `x=y` over `x = y` here.

File: tools/stack_decode.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+      if stackaddr_match:
+        thread_id, frame_idx, address = stackaddr_match.groups()
+        traces[thread_id].address_list.append((frame_idx, address))
+        continue
+      trace_end_match = trace_end_re.search(line)
+      if trace_end_match:
+        thread_id = trace_end_match.groups()[0]
+        output_stacktrace(thread_id, traces[thread_id])
+      else:
+        print line, # Pass through print all other log lines
```
Comment (htuch) [inline tools/stack_decode.py]: Nit: , at EOL

File: tools/stack_decode.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+  except KeyboardInterrupt:
+    return
+
+# Output one stacktrace after passing it through addr2line with appropriate
+# options
+def output_stacktrace(thread_id, traceinfo):
+  log_prefix, object_file, stack = traceinfo
+  piped_input = ""
```
Comment (htuch) [inline tools/stack_decode.py]: Nit: should be consistent about quoting style in Python. Pick either '' or "", I have a personal preference for '', but it doesn't matter as long as it's consistent.

File: tools/stack_decode.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+  for stack_frame in stack:
+    piped_input += (stack_frame[1] + '\n')
```
Comment (htuch) [inline tools/stack_decode.py]: Best to document why you start at this stack frame.

File: source/server/server.cc (hunk 3 above)
Comment (htuch) [inline source/server/server.cc]: As we came to the conclusion IRL, we want the backward dependency to be in only a few files that can be easily excluded from the build. Maybe we can add an abstract `backtrace` interface to `include/envoy` and have folks only depend on that. This would include APIs to dump stacktrace. Then, in `envoy_build_system.bzl`, when setting up the `deps` for things like `envoy_cc_library`, we can point it at the specific implementation that uses Backwards. For Google's (or anyone else who wants to override backwards) import, they can replace this dependency with their own.

@mattklein123 How do you override the signal handlers in the Lyft build? We would like to be able to wrap `main()` with our own extras for initialization (and others might want to as well).

File: source/common/common/assert.h [hunk 13]
```diff
@@ -25,6 +27,7 @@
  * Indicate a panic situation and exit.
  */
 #define PANIC(X)                                                                                   \
+  BACKTRACE_PROD_LOG();                                                                            \
```
Comment (htuch) [inline source/common/common/assert.h]: Maybe remove these from this PR and wait until the signal handler PR one so that we can just trigger on `sigabrt`.

File: source/server/server.cc (hunk 3 above)
Comment (mattklein123) [inline source/server/server.cc]: We don't currently, we just use the built-in libevent signal processing functionality for SIGTERM, SIGHUP, etc. We let Envoy crash and dump and then have an external process which sucks up the core dump and does stuff with it.

I think it makes sense to document which signals Envoy itself handles and then just allow people to override things like SIGABRT, SIGSEGV, etc. in main if they want. I'm open to whatever solution here makes the most sense. We could have a stub signal setup library that is called from main that is empty in the OSS build, and you could override?

Also, while we are on this topic, my vote would be to make the backtrace stuff opt-in on production builds, and default on debug builds.

File: tools/stack_decode.py (hunk 12 above)
Comment (dnoe) [inline tools/stack_decode.py]: This index operation is actually picking the address out of the tuple (first element captured is the thread ID).  I'm going to change this slightly to have better clarity.

File: source/server/server.cc (hunk 3 above)
Comment (htuch) [inline source/server/server.cc]: I think we should refactor `main.cc` into site and non-site specific files, where the site-specific file is where the `main` symbol lives. It will be necessary for some site to have additional flag, logging, core dump, signal handler, memory management and other deployment specific init code that might want to run before we jump into Envoy proper (and teardown code on the way out). This can be largely opaque to the regular OSS build as long as we don't put too much code inside `main()` and can replace it with our own.

Comment (mattklein123) [inline source/server/server.cc]: Yup that works for me. (Basically what you did for test main).

File: tools/stack_decode.py (hunk 9 above)
Comment (dnoe) [inline tools/stack_decode.py]: Yes, it did (and also likes " instead of ')

File: bazel/envoy_build_system.bzl [hunk 14]
```diff
@@ -17,16 +17,12 @@ def envoy_copts(repository):
...
-        repository + "//bazel:opt_build": [],
+        repository + "//bazel:opt_build": ["-ggdb3"],
         repository + "//bazel:fastbuild_build": [],
         repository + "//bazel:dbg_build": ["-ggdb3"],
     }) + select({
         repository + "//bazel:disable_tcmalloc": [],
         "//conditions:default": ["-DTCMALLOC"],
-    }) + select({
-        # Allow debug symbols to be added to opt/fastbuild as well.
-        repository + "//bazel:debug_symbols": ["-ggdb3"],
```
Comment (htuch) [inline bazel/envoy_build_system.bzl]: Can you also remove `debug_symbols` from `bazel/BUILD`?

File: tools/stack_decode.py [hunk 15]
```diff
@@ -17,63 +17,72 @@
...
+    piped_input += (stack_addr + "\n")
+  addr2line = subprocess.Popen(
+      ["addr2line", "-Cpisfe", traceinfo.obj_file],
+      stdin=subprocess.PIPE,
+      stdout=subprocess.PIPE)
   output_stdout, _ = addr2line.communicate(piped_input)
-  output_lines = output_stdout.split('\n')
+  output_lines = output_stdout.split("\n")
 
   resolved_stack_frames = list(enumerate(output_lines, start=1))
```
Comment (htuch) [inline tools/stack_decode.py]: Nit: shouldn't need to convert to list, leaving as an iterator is fine (and more efficient, not that it matters here) as it is consumed only in the `for` loop below.

File: tools/stack_decode.py [hunk 16]
```diff
@@ -17,63 +17,72 @@
...
-  print "%s Backtrace (most recent call first) from thread %s:" % (log_prefix, thread_id)
+  print "%s Backtrace (most recent call first) from thread %s:" % (
+      traceinfo.log_prefix, thread_id)
   for stack_frame in resolved_stack_frames:
     print "  #%s %s" % (stack_frame[0], stack_frame[1])
```
Comment (htuch) [inline tools/stack_decode.py]: Nit: this can just be written `% stack_frame`.

File: tools/stack_decode.py [hunk 17]
```diff
@@ -17,63 +17,72 @@
...
+        traces[thread_id].address_list.append(address)
         continue
       trace_end_match = trace_end_re.search(line)
       if trace_end_match:
         thread_id = trace_end_match.groups()[0]
         output_stacktrace(thread_id, traces[thread_id])
       else:
-        print line, # Pass through print all other log lines
+        # Pass through print all other log lines:
+        print line,
```
Comment (htuch) [inline tools/stack_decode.py]: Nit: still the trailing ,

Comment (dnoe) [inline tools/stack_decode.py]: Trailing , inhibits print from adding the extra newline, it's deliberate.

File: bazel/envoy_build_system.bzl (hunk 14 above)
Comment (dnoe) [inline bazel/envoy_build_system.bzl]: Sure

File: tools/stack_decode.py (hunk 17 above)
Comment (htuch) [inline tools/stack_decode.py]: I learnt something new about Python today.

File: tools/stack_decode.py [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+# In each case this script will decode any backtrace log lines found and echo
+# back all non-Backtrace lines untouched.
+
+import collections
+import re
+import subprocess
+import sys
+
+Backtrace = collections.namedtuple("Backtrace",
+                                   "log_prefix obj_file, address_list")
```
Comment (mattklein123) [inline tools/stack_decode.py]: nit: any reason why there isn't a comma after log_prefix, or switching to whitespace between obj_file and address_list? From docs it looks like both are valid but seems odd to use both?

Comment (dnoe) [inline tools/stack_decode.py]: No reason, I think it came from my originally using commas and pyformat tool mangling it.  I'll push one more fix for this.

**Comment by mattklein123** [comment]
@dnoe also needs a master merge.

**Comment by ldemailly** [comment]
complete envoy noob question here but : is there a build mode or runtime flag of envoy to not try to trap crashes at all and just fall back to generating a core (or even better a wrapper that gdb "bt all" after crash)

**Comment by mattklein123** [comment]
https://github.com/lyft/envoy/tree/master/bazel#stack-trace-symbol-resolution
